### PR TITLE
[FLINK-26421] Use only EnvironmentSettings to configure the environment

### PR DESCRIPTION
**Pull request by matriv**
## What is the purpose of the change

Use `EnvironmentSettings` with the new method `withConfiguration` in its
`Builder` to specify configuration options on top of the one inherited by the
environment (flink-conf.yml, CLI params).

## Brief change log

  - Remove planner & executor string identifiers
  - Remove deprecated create methods which use `TableConfig` instead of `EnvironmentSettings`.
  - Add an underlying `Configuration` in `EnvironmentSettings` so that they contain any extra use defined config options.
  - The `TableConfig` now implements `ReadableConfig and contains all of the config options specified by the env (`flink-conf.yml`, `CLI` params), in the underlying `rootConfiguration`, plus all the extra configuration defined by the user app through the `EnvironmentSettings`. 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): **no**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: **no**
  - The serializers: **no**
  - The runtime per-record code paths (performance sensitive): **no**
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: **no**
  - The S3 file system connector: **no**

## Documentation

  - Does this pull request introduce a new feature? **yes**
  - If yes, how is the feature documented? **docs / JavaDocs**

### CODE REVIEW

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/harness/HarnessTestBase.scala [hunk 1]
```diff
@@ -121,7 +124,7 @@ class HarnessTestBase(mode: StateBackendMode) extends StreamingTestBase {
     elements.filter(e => !e.isInstanceOf[Watermark]).toList
   }
 
-  class TestTableConfig extends TableConfig {
+  class TestTableConfig() extends TableConfig {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/harness/HarnessTestBase.scala:127]: unnecessary change?

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/harness/TableAggregateHarnessTest.scala [hunk 2]
```diff
@@ -49,8 +49,8 @@ class TableAggregateHarnessTest(mode: StateBackendMode) extends HarnessTestBase(
   override def before(): Unit = {
     super.before()
     val setting = EnvironmentSettings.newInstance().inStreamingMode().build()
-    val config = new TestTableConfig
-    this.tEnv = StreamTableEnvironmentImpl.create(env, setting, config)
+//    val config = new TestTableConfig
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/harness/TableAggregateHarnessTest.scala:52]: remove?

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for the PR @matriv. I like how minimal the changes are. It would be great to add some additional tests that verify:

- A configuration option set in `StreamExecutionEnvironment` is available from `TableConfig.get` in `StreamTableEnvironment`.
- A configuration option set in `EnvironmentSettings` can influence the `StreamExecutionEnvironment#getConfiguration` in `TableEnvironment`.
- A configuration option set in `TableConfig` has no influence on `StreamExecutionEnvironment` if it is a deployment option.

File: flink-clients/src/main/java/org/apache/flink/client/program/StreamContextEnvironment.java [hunk 3]
```diff
@@ -185,6 +184,7 @@ private JobExecutionResult getJobExecutionResult(final JobClient jobClient) thro
     @Override
     public JobClient executeAsync(StreamGraph streamGraph) throws Exception {
         validateAllowedExecution();
+        checkNotAllowedConfigurations();
```
[the remark this answers is not in the document]
Comment (matriv) [reply flink-clients/src/main/java/org/apache/flink/client/program/StreamContextEnvironment.java:187]: I was going to, thx, just kept it there to test everything.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java [hunk 4]
```diff
@@ -311,6 +311,7 @@ public ExecutionConfig getConfig() {
      */
     public StreamExecutionEnvironment setParallelism(int parallelism) {
         config.setParallelism(parallelism);
+        configuration.set(CoreOptions.DEFAULT_PARALLELISM, parallelism);
```
[the remark this answers is not in the document]
Comment (matriv) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java:314]: Sorry, leftover

File: flink-table/flink-table-api-scala-bridge/src/main/scala/org/apache/flink/table/api/bridge/scala/internal/StreamTableEnvironmentImpl.scala [hunk 5]
```diff
@@ -292,6 +294,12 @@ class StreamTableEnvironmentImpl (
 
 object StreamTableEnvironmentImpl {
 
+  def create(
+              executionEnvironment: StreamExecutionEnvironment,
+              settings: EnvironmentSettings): StreamTableEnvironmentImpl =
+    create(executionEnvironment, settings, TableConfig.getDefault)
+
+  @VisibleForTesting
```
[the remark this answers is not in the document]
Comment (matriv) [reply flink-table/flink-table-api-scala-bridge/src/main/scala/org/apache/flink/table/api/bridge/scala/internal/StreamTableEnvironmentImpl.scala:302]: It's only used by `OverAggregateHarnessTest` because of the "weird" testing with a custom `TestTableConfig` class to explicitly set `maxIdleStateRetentionTime`, overriding the default calculation on the original `TableConfig`. Haven't thought of a workaround to avoid this, any suggestions?

[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-api-scala-bridge/src/main/scala/org/apache/flink/table/api/bridge/scala/internal/StreamTableEnvironmentImpl.scala:302]: Let's rework the test then. I don't see a reason why we should make our internal methods complicated just for a wrongly implemented test.

**Comment by matriv** [comment]
@twalthr Please check, I hope I've addressed all the comments.

**Comment by matriv** [comment]
@flinkbot run azure

**Comment by matriv** [comment]
@flinkbot run azure

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/EnvironmentSettings.java [hunk 6]
```diff
@@ -60,44 +57,21 @@
...
-        this.planner = planner;
-        this.executor = executor;
-        this.builtInCatalogName = builtInCatalogName;
-        this.builtInDatabaseName = builtInDatabaseName;
-        this.isStreamingMode = isStreamingMode;
+    private final Configuration configuration;
+
+    private EnvironmentSettings(Configuration configuration) {
+        this.configuration = configuration;
+        if (configuration.get(RUNTIME_MODE) == AUTOMATIC) {
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/EnvironmentSettings.java:68]: I'm wondering if this location is too eager to check. Can you check one more time what happens if the runtime mode is not set in environment settings and is AUTOMATIC in `StreamExecutionEnviornment`. What is the behavior of `StreamTableEnviornment.create(env)`?

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/harness/HarnessTestBase.scala [hunk 7]
```diff
@@ -120,23 +118,6 @@ class HarnessTestBase(mode: StateBackendMode) extends StreamingTestBase {
   def dropWatermarks(elements: Array[AnyRef]): util.Collection[AnyRef] = {
     elements.filter(e => !e.isInstanceOf[Watermark]).toList
   }
-
-  class TestTableConfig extends TableConfig {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/harness/HarnessTestBase.scala:124]: can we make `TableConfig` class final now?

**Review by twalthr — CHANGES_REQUESTED** [review]
Thanks for the update @matriv. I added my last set of comments. Should be good in the next iteration.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/context/ExecutionContext.java [hunk 8]
```diff
@@ -99,18 +99,20 @@ public StreamTableEnvironment getTableEnvironment() {
 
     private StreamTableEnvironment createTableEnvironment() {
         // checks the value of RUNTIME_MODE
```
[the remark this answers is not in the document]
Comment (matriv) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/context/ExecutionContext.java:101]: Yep, it's being checked in the `build() -> new EnvironmentSettings(Configuration config)`

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/EnvironmentSettings.java (hunk 6 above)
Comment (matriv) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/EnvironmentSettings.java:68]:
```
Exception in thread "main" org.apache.flink.table.api.TableException: Unknown runtime mode 'AUTOMATIC'. This is a bug. Please consider filing an issue.
	at org.apache.flink.table.planner.delegation.DefaultPlannerFactory.create(DefaultPlannerFactory.java:72)
	at org.apache.flink.table.factories.PlannerFactoryUtil.createPlanner(PlannerFactoryUtil.java:52)
	at org.apache.flink.table.api.bridge.java.internal.StreamTableEnvironmentImpl.create(StreamTableEnvironmentImpl.java:124)
	at org.apache.flink.table.api.bridge.java.StreamTableEnvironment.create(StreamTableEnvironment.java:122)
	at org.apache.flink.table.api.bridge.java.StreamTableEnvironment.create(StreamTableEnvironment.java:94)
	at org.apache.flink.table.examples.java.basics.StreamSQLExample.main(StreamSQLExample.java:55)
```

Comment (matriv) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/EnvironmentSettings.java:68]: I believe we need to keep it.

Comment (twalthr) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/EnvironmentSettings.java:68]: This exception message is not correct. This is not a bug but intended behavior for some unsupported feature. I would suggest to movee the check from the constructor of `EnvironmentSettings` to `DefaultPlannerFactory.create`. Or have the same message at both locations.

**Review by twalthr — APPROVED** [review]
LGTM if build is green

**Comment by matriv** [comment]
@twalthr Thx a lot for the improved java docs! Looks good!
